(PUP-7732) Acceptance audit

johnduarte opened the pull request.

puppetcla commented:
CLA signed by all contributors.

justinstoller commented:
@branan @er0ck This would be awesome to get looked at soon

branan commented:
I'll defer to QA on this.

er0ck commented:
👀 on

er0ck commented on acceptance/tests/allow_arbitrary_node_name_for_apply.rb:
```diff
@@ -1,5 +1,9 @@
 
+tag 'audit:medium',
+    'audit:integration',  # Ruby level integration tests already exist. This acceptance test can be deleted.
+    'audit:delete'
```
any reason this PR shouldn't just delete this test?  i had thought about using an audit:delete tag as well.  i think our progress tooling can count the number of tests and display that near the job runtime

johnduarte replied:
I believe the audit should be restricted to presenting its recommendation and not take any action upon those recommendations. The delete action should be part of the implementation effort of the audit recommendations.

er0ck replied:
i'm fine either way. i'll only add that i think people want to see and be able to show results.  soon.

er0ck commented on acceptance/tests/apply/augeas/hosts.rb:
```diff
@@ -1,8 +1,10 @@
-skip_test 'requires augeas which is included in AIO' if @options[:type] != 'aio'
+tag 'risk:medium',
+    'audit:medium',
+    'audit:acceptance'
```
suggest:
```
audit:refactor' # move to type directory 
# this is not testing puppet apply
```

er0ck replied:
suggest:
```
'audit:refactor' # move to puppet types test directory, this is not testing puppet apply
# reduce to a single manifest and apply
```

johnduarte replied:
:+1:

er0ck commented on acceptance/tests/apply/augeas/puppet.rb:
```diff
@@ -1,8 +1,10 @@
-  skip_test 'requires augeas which is included in AIO' if @options[:type] != 'aio'
+  tag 'risk:medium',
+      'audit:medium',
+      'audit:acceptance'
```
suggest:
```
'audit:refactor' # move to types test dir
# remove spurious `puppet config set`
```

johnduarte replied:
:+1:

er0ck commented on acceptance/tests/apply/augeas/services.rb:
```diff
@@ -1,8 +1,10 @@
-  skip_test 'requires augeas which is included in AIO' if @options[:type] != 'aio'
+  tag 'risk:medium',
+      'audit:medium',
+      'audit:acceptance'
```
suggest:
```
'audit:refactor' # move to types test dir
# use single manifest/apply
```

johnduarte replied:
:+1:

er0ck commented on acceptance/tests/apply/classes/parameterized_classes.rb:
```diff
@@ -1,5 +1,9 @@
 
+tag 'audit:high',
+    'audit:unit',  # This should be covered at the unit layer.
+    'audit:delete'
```
is it covered at unit?  if so, can this be deleted in this PR?

johnduarte replied:
This is a spurious use of the `audit:delete` tag. Thanks for pointing it out. I will remove it and the others as noted.

er0ck commented on acceptance/tests/apply/classes/should_allow_param_override.rb:
```diff
@@ -1,5 +1,9 @@
 
+tag 'audit:high',
+    'audit:unit',  # This should be covered at the unit layer.
+    'audit:delete'
```
is it covered at unit? if so, can this be deleted in this PR?

er0ck commented on acceptance/tests/apply/classes/should_allow_param_undef_override.rb:
```diff
@@ -1,5 +1,10 @@
 
+tag 'audit:high',
+    'audit:unit',  # This should be covered at the unit layer.
+    'audit:delete'
```
is it covered at unit? if so, can this be deleted in this PR?

spurious newline

er0ck commented on acceptance/tests/agent/agent_disable_lockfile.rb:
```diff
@@ -1,6 +1,12 @@
 
+tag 'audit:integration', # lockfile uses the standard `vardir` location to store/query lockfile.
+                         # The validation of the `vardir` at the OS level
+                         # should be accomplished in another test.
```
is this implying that we can test both vardir/lockfile location as well as being able to disable the lockfile in integration layer?

johnduarte replied:
So long as the management of the `vardir` is validated at the OS level, the ability to manage the lockfile within that location can be covered at the integration level.

er0ck replied:
as long as we have an agent acceptance that proves the presence of the lockfile prevents an agent run correctly.

er0ck commented on acceptance/tests/apply/puppet_apply_graph.rb:
```diff
@@ -1,5 +1,9 @@
 test_name 'puppet apply should generate a graph'
 
+tag 'audit:low',
+    'audit:integration',  # Core testing of `vardir` should occur in another test.
```
is this implying that we can test the graph output in the integration layer?

johnduarte replied:
This test does not test the graph output only that a file is created. Such a validation can be covered at the integration level.

er0ck replied:
or maybe the test should ensure the graph output?

er0ck replied:
same here.  i'll add this to testrail and can merge this i think

er0ck commented on acceptance/tests/concurrency/ticket_2659_concurrent_catalog_requests.rb:
```diff
@@ -1,5 +1,9 @@
 
+tag 'audit:low',
+    'audit:integration',  # Testing server
+    'server'
```
is that comment in the wrong place?  or can it be an integration test because we're 'testing server'?

i agree the catalog endpoint being able to handle concurrent requests should be tested in its integration tests, and while it is busy, as here.

johnduarte replied:
The comment is intended to indication that the test should be in the puppetserver integration tests. This information is covered by the tags themselves. I will delete the comment.

er0ck commented on acceptance/tests/config/apply_file_metadata_specified_in_config.rb:
```diff
@@ -1,5 +1,8 @@
 test_name "#17371 file metadata specified in puppet.conf needs to be applied"
```
the ticket # here is a bit confusing.  should we remove or link to (and create) a testrail entry?

er0ck commented on acceptance/tests/direct_puppet/catalog_uuid_correlates_catalogs_with_reports.rb:
```diff
@@ -1,4 +1,9 @@
+
+  tag 'audit:medium',
+      'audit:acceptance',
+      'server'
```
doesn't this just test that the uuid is the same in catalog and report when running the agent with --use-cached-catalog?

great abstractions in this test make it very easy to read!

johnduarte replied:
Yes, and in doing so perform an end-to-end validation that cached catalogs behave properly and retain the expected uuid between the server report and the applied agent catalog. Does this not map to a server based acceptance test?

er0ck replied:
seems like an agent test to me.  it doesn't need the server if we can give it a catalog fixture.

justinstoller commented:
@er0ck requested that this be merged into 5.0.x and then merged up into master. I've retargeted the PR accordingly.

justinstoller commented on acceptance/tests/reports/corrective_change_via_puppet.rb:
```diff
@@ -10,6 +10,17 @@
+  tmp_environment = mk_tmp_environment_with_teardown(master, test_file_name)
+  tmp_file = {}
+
+>>>>>>> (PUP-7732) Audit Acceptance Tests - Reports
```
We probably shouldn't merge this

johnduarte commented:
@er0ck I believe I have addressed all of your commented concerns and updated testrail cases as appropriate. Please let me know if there is anything I have missed.

er0ck commented:
i'm fine merging this against master.  there is a lot of work going on in 5.0.x so we should get as many test changes in there as possible.  i suggest executing this audit against 5.0.x or cherry picking these commits back there.
